### PR TITLE
Match our isupper with python behavior

### PR DESCRIPTION
**Pull request by grdanny**
As mentioned in:

https://github.com/imageworks/pystring/issues/45

Our **isupper** relies heavily on the more strict isupper from the standard library.
So something like this gets true in python, but false in c++
```
>>> "A1".isupper()
True
```
This should make our behavior match the python one and add some tests so we won't regres.

### CODE REVIEW

**Comment by lgritz** [comment]
So python isupper means "nothing is a lower case letter"? But things that aren't "letters" are ignored from the decision?

Are there other pystring library calls that also need to be adjusted in similar ways?

**Comment by grdanny** [comment]
Looks like the user that reported this learned this the hard way somehow.
It's pretty subtle and counter intuitive, but yeah, "HELLO 123" is considred upper by python.
If you asked me off the top of my head I would probably also answer false.

I'm not sure if we are missing any other subtleties like that. 
But it wouldn't be too hard to write something in python that throws all sort of examples and edge cases at the corresponding python3 functions and then exports out the results to some json file or something and we read that in c++ and make sure it would match.

It can be an additional unit test or something... and can maybe flush out a few more cases like this we never thought about?

**Comment by grdanny** [comment]
hmm... just playing around with islower I can see we are missing some cases there too...
for example:

in python:
```
>>> "hello!".islower()
True
```
but ours will return false.
I'll look around for some more and expand this PR a bit.

**Comment by grdanny** [comment]
I had claude generate a bunch of python code that uses all the functions with different cases and returns their values.
And then I just moved those to our test.cpp with the expected values.

Overall we were in good shape, but it uncovered a couple of more things so there are additional small fixes.

File: pystring_impl.h [hunk 1]
```diff
@@ -927,6 +941,16 @@ const std::string colon = ":";
         int nummatches = 0;
         int cursor = start;
 
+        // special handling for an empty substring
+        // this will match python's behavior of
+        // "bob".count("") == 4
+        // "".count("") == 1
```
Comment (lgritz) [inline pystring_impl.h:947]: Wait, what? How is `"bob".count("")` equal to 4?

File: test.cpp [hunk 2]
```diff
@@ -7,6 +7,25 @@
 #include "pystring.h"
 #include "unittest.h"
 
+// Helper wrappers (for pystring functions that don't have direct return values)
+
+namespace {
+static std::vector<std::string> pystring_split(const std::string& s, int maxsplit)
```
Comment (lgritz) [inline test.cpp:13]: Why is this static? Maybe `inline`?

Comment (lgritz) [reply test.cpp:13]: never mind, I see that this is just in test.cpp.

**Review by lgritz — APPROVED** [review]

File: pystring_impl.h (hunk 1 above)
Comment (grdanny) [reply pystring_impl.h:947]: weird. I know, but that's the python behavior is:
```
"bob".count("")  
4
```
I guess it's counting the non strings between the letters?

It's double weird because:

```
>>> "bob".split("")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: empty separator
```

So it's a little inconsistent, but that's the behavior.

Comment (lgritz) [reply pystring_impl.h:947]: Brain melting. I looked at "bob" and could have sworn that there were at least two "" between each letter. Maybe three if you packed them in carefully!
